datavault-webapp/src/main/java/org/datavaultplatform/webapp/security/ScopedPermissionEvaluator.java

```diff
@@ -0,0 +1,112 @@
+package org.datavaultplatform.webapp.security;
+
+import com.google.common.collect.Multimap;
+import org.datavaultplatform.common.model.Group;
+import org.datavaultplatform.common.model.Permission;
+import org.datavaultplatform.common.model.Vault;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.security.access.PermissionEvaluator;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.TreeMap;
+
+/**
+ * A permission expression evaluator that is able to check existence of {@link Vault} and {@link Group} permissions
+ * on a given authentication.
+ *
+ * <p>
+ * For example, it can be used in a pre-authorize filter:
+ * <pre>
+ *    {@literal @}PreAuthorize("hasPermission(#vault, 'VAULT_PERMISSION_1')")}
+ *     public void deleteVault(Vault vault) {
+ *         ...
+ *     }
+ * </pre>
+ *
+ * <p>
+ * Or called imperatively using any {@link Permission} value.
+ * <pre>
+ *     Vault vault = ...;
+ *     if (permissionEvaluator.hasPermission(user, vault, Permission.VAULT_PERMISSION_1)) { ... }
+ * </pre>
+ */
+public class ScopedPermissionEvaluator implements PermissionEvaluator {
+
+    private static final Logger logger = LoggerFactory.getLogger(ScopedPermissionEvaluator.class);
+
+    /**
+     * Real implementation for actual checking of permissions on an authentication.
+     *
+     * @param auth         The authentication to extract permissions from.
+     * @param key          The unique key representing the scope of the requested permission.
+     * @param permissionId A string or enum value representing the permission.
+     * @return {@code true} iff authorized, {@code false} otherwise.
+     */
+    boolean hasPermission(Authentication auth, ScopedPermissionKey key, Object permissionId) {
+        if (auth == null) {
+            return false;
+        }
+
+        Collection<? extends GrantedAuthority> authorities = auth.getAuthorities();
+        if (authorities.contains(new SimpleGrantedAuthority("ROLE_ADMIN"))) {
+            return true;
+        }
+
+        Permission permission = null;
+
+        if (permissionId instanceof String) {
+            try {
+                permission = Permission.valueOf(((String) permissionId).toUpperCase());
+            } catch (IllegalArgumentException ex) {
+                logger.warn("Attempted to check a non-existent permission ({}, {})", key, permissionId);
+            }
+        } else if (permissionId instanceof Permission) {
+            permission = (Permission) permissionId;
+        }
```

Comment:
Probably worth logging an error (or maybe even throwing) if we don't get into either clause here

Reply:
Throwing makes more sense instead of silently failing I think. Will update.